src/lib.ml

```diff
@@ -1337,24 +1304,87 @@ module DB = struct
         | Some (loc, implements), Some variant ->
           if not (Lib_name.Map.mem libmap implements) then
             Errors.fail loc
-              "Library implementation %a for variant %a implements a library \
-               outside the project. This is forbidden."
-              Lib_name.pp implements Variant.pp variant
+              "Library implementation %a for variant %a implements a library @ \
+               outside the project.@ Instead of using (variant %a) here,@ you \
+               need to reference it in the virtual library definition, using @ \
+               (external_implementations (%a %a))."
+              Lib_name.pp implements Variant.pp variant Variant.pp variant
+              Variant.pp variant Lib_name.pp lib.name
+
         | _, _ -> ()
         end
       | Redirect (_, _) (* skip b/c [Found] covers *) -> ()
       | Hidden (_, _) -> assert false
       | Not_found -> assert false)
 
-  let create_from_library_stanzas ?parent ~lib_config stanzas =
+  let extract_loc_from_variant_map vmap =
+    Variant.Map.choose vmap
+    |> Option.map ~f:(fun (_, (loc, _)) -> loc)
+
+  let check_valid_external_variants
+        (libmap : resolve_result Lib_name.Map.t)
+        (variant_map : (Loc.t * Lib_name.t) Variant.Map.t Lib_name.Map.t) =
+    Lib_name.Map.iteri variant_map ~f:(fun vlib variants ->
+      let loc = extract_loc_from_variant_map variants in
+      match Lib_name.Map.find libmap vlib with
+      | None ->
+        Errors.fail_opt loc
+          "Virtual library %a hasn't been found in the project."
+          Lib_name.pp vlib
+      | Some (Found (lib : Lib_info.t)) when Option.is_none lib.virtual_ ->
+        Errors.fail_opt loc
+          "Library %a isn't a virtual library."
+          Lib_name.pp vlib
+      | Some (Redirect (None, best_name)) ->
+        Errors.fail_opt loc
+          "To declare variants you must use %a's public name,@ which is %a."
```

Comment:
Why isn't it allowed to use a private name?

Reply:
Well this is my bad, this limitation shouldn't exist we can simply convert private names to the corresponding public name. I'll update the code